src/main/java/org/apache/flink/connector/clickhouse/internal/executor/ClickHouseUpsertExecutor.java

```diff
@@ -96,12 +116,14 @@ public void setRuntimeContext(RuntimeContext context) {}
     public void addToBatch(RowData record) throws SQLException {
         switch (record.getRowKind()) {
             case INSERT:
-                insertConverter.toExternal(record, insertStmt);
-                insertStmt.addBatch();
-                break;
             case UPDATE_AFTER:
-                updateConverter.toExternal(updateExtractor.apply(record), updateStmt);
-                updateStmt.addBatch();
+                if (exist(record)) {
```

Comment:
There is no need to call exist before insert rowData, causing performance loss.

Reply:
I agree performance loss when call exist for each row.
1.When user specify upsert mode if a primary key is defined on the DDL, we need check row exist before insert/update.
2.Otherwise, we don't need call exist statement, casing performance loss.